x/migration/keeper/msg_server_claim_morse_application.go

```diff
@@ -2,17 +2,142 @@ package keeper
 
 import (
 	"context"
+	"strings"
 
-	sdk "github.com/cosmos/cosmos-sdk/types"
+	cosmostypes "github.com/cosmos/cosmos-sdk/types"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
 
+	"github.com/pokt-network/poktroll/app/volatile"
+	apptypes "github.com/pokt-network/poktroll/x/application/types"
 	migrationtypes "github.com/pokt-network/poktroll/x/migration/types"
+	sharedtypes "github.com/pokt-network/poktroll/x/shared/types"
 )
 
-func (k msgServer) ClaimMorseApplication(goCtx context.Context, msg *migrationtypes.MsgClaimMorseApplication) (*migrationtypes.MsgClaimMorseApplicationResponse, error) {
-	ctx := sdk.UnwrapSDKContext(goCtx)
+func (k msgServer) ClaimMorseApplication(ctx context.Context, msg *migrationtypes.MsgClaimMorseApplication) (*migrationtypes.MsgClaimMorseApplicationResponse, error) {
+	sdkCtx := cosmostypes.UnwrapSDKContext(ctx)
+	logger := k.Logger().With("method", "ClaimMorseApplication")
 
-	// TODO: Handling the message
-	_ = ctx
+	if err := msg.ValidateBasic(); err != nil {
+		return nil, status.Error(codes.InvalidArgument, err.Error())
+	}
 
-	return &migrationtypes.MsgClaimMorseApplicationResponse{}, nil
+	// DEV_NOTE: It is safe to use MustAccAddressFromBech32 here because the
+	// shannonDestAddress is validated in MsgClaimMorseApplication#ValidateBasic().
+	shannonAccAddr := cosmostypes.MustAccAddressFromBech32(msg.ShannonDestAddress)
+
+	// Ensure that a MorseClaimableAccount exists for the given morseSrcAddress.
+	morseClaimableAccount, isFound := k.GetMorseClaimableAccount(
+		sdkCtx,
+		msg.MorseSrcAddress,
+	)
+	if !isFound {
+		return nil, status.Error(
+			codes.NotFound,
+			migrationtypes.ErrMorseApplicationClaim.Wrapf(
+				"no morse claimable account exists with address %q",
+				msg.MorseSrcAddress,
+			).Error(),
+		)
+	}
+
+	// Ensure that the given MorseClaimableAccount has not already been claimed.
+	if morseClaimableAccount.IsClaimed() {
+		return nil, status.Error(
+			codes.FailedPrecondition,
+			migrationtypes.ErrMorseApplicationClaim.Wrapf(
+				"morse address %q has already been claimed at height %d by shannon address %q",
+				msg.MorseSrcAddress,
+				morseClaimableAccount.ClaimedAtHeight,
+				morseClaimableAccount.ShannonDestAddress,
+			).Error(),
+		)
+	}
+
+	// Default to the application stake amount recorded in the MorseClaimableAccount.
```

Comment:
Per the comments in the other PRs and below.

1. This cannot be nil
2. claimMsg.Stake must equal onchainMorseState.Stake

Reply:
see https://github.com/pokt-network/poktroll/pull/1082#discussion_r1980898059